### PR TITLE
For #28450 - Fixes to the context comparison code and HumanUser name stored in the path cache

### PR DESCRIPTION
**Pull request by alandann**
- Context comparisons now only check type and id for all entity dictionaries where previously the entire dictionary would have been compared which could lead to unexpected comparison failures.
- HumanUser.name is now stored in the path cache for user sandbox entities instead of HumanUser.login - this fixes an issue where a context created from a user sandbox path would end up with the login instead of the name in context.user["name"]

### CODE REVIEW

File: python/tank/context.py [hunk 1]
```diff
@@ -113,23 +113,78 @@ def __str__(self):
...
+        
+        return True 
+    
+    def _entity_dicts_eq(self, d1, d2):
+        """
+        Test to see if two entity dictionaries are equal.  They are considered
+        equal if both are dictionaries containing 'type' and 'id' with the same
+        values for both keys, e.g.:
+        
+        {"type":"Shot", "id":123, "foo":"foo"} == {"type":"Shot", "id":123, "bar":"bar"}
```
Comment (manneohrstrom) [inline python/tank/context.py]: Would be nice if the comment stated the result of the example comparison more clearly I think, e.g.

```
Comparing the following two dictionaries would return True:
- {"type":"Shot", "id":123, "foo":"foo"}
- {"type":"Shot", "id":123, "bar":"bar", "other_field": "baz"}
```
Also threw in an "other field" in the example to show that it's okay if the dicts contain other unrelated fields too. What do you think?

File: python/tank/context.py [hunk 2]
```diff
@@ -113,23 +113,78 @@ def __str__(self):
...
+        """
+        Test to see if two entity dictionaries are equal.  They are considered
+        equal if both are dictionaries containing 'type' and 'id' with the same
+        values for both keys, e.g.:
+        
+        {"type":"Shot", "id":123, "foo":"foo"} == {"type":"Shot", "id":123, "bar":"bar"} 
+
+        :param d1:  First entity dictionary
+        :param d2:  Second entity dictionary
+        :returns:   True of d1 == d2, False otherwise
```
Comment (manneohrstrom) [inline python/tank/context.py]: Sorry, this is such minor comments, but I find the use of `==` here a little confusing. Would it be clearer if it said something like 'true if id and type keys are the same for d1 and d2, false otherwise"?

File: python/tank/context.py [hunk 3]
```diff
@@ -113,23 +113,78 @@ def __str__(self):
...
+        if not self._entity_dicts_eq(self.entity, other.entity):
+            return False
+        
+        if not self._entity_dicts_eq(self.step, other.step):
+            return False
+        
+        if not self._entity_dicts_eq(self.task, other.task):
+            return False
+        
+        if not self._entity_dicts_eq(self.user, other.user):
```
Comment (manneohrstrom) [inline python/tank/context.py]: given that user currently may result in a performance hit, should this happen at the very end?

**Comment by manneohrstrom** [comment]
Looks sweet! Chatted to Alan off line about whether to put `_entity_dicts_eq` as a internal method inside `__eq__` or not - perhaps just as a way to indicate that this method is just used for this comparison and shouldn't really be reused for other stuff. 

The only other suggestion I have is that we should add a unit test which checks when the id and type are the same but when the names are off (e.g. this particular bug). It looks like there were other unit tests in place for basic checks, but not for edge cases like that!

Alan, back to you - now you know my feedback please make changes as you see fit and relevant.

**Comment by benoit-leveau** [comment]
Hi guys,

Is it okay to try it as is and see if it fixes our particular issue?
(ie. "*this
fixes an issue where a context created from a user sandbox path would end
up with the login instead of the name in context.user["name"]*")

Thanks for looking into this.

Cheers,
Benoit

File: python/tank/context.py (hunk 1 above)
Comment (jfboismenu-adsk) [inline python/tank/context.py]: I concur.

File: python/tank/context.py [hunk 4]
```diff
@@ -113,23 +113,78 @@ def __str__(self):
...
+            return False
+
+        # compare additional entities - note that currently this assumes the additional entity 
+        # lists are in exactly the same order for both self and other.        
+        if isinstance(self.additional_entities, list) and isinstance(other.additional_entities, list):
```
Comment (jfboismenu-adsk) [inline python/tank/context.py]: From my understanding, additional_entities should always be lists based on all the code that depends on it everywhere.

If having something other than a list can break the internal logic of the class and all its clients, I suggest you validate that additional_entities is an array in __init__ and raise an exception there instead of having context comparison fail because of bad input.

**Comment by alandann** [comment]
Hi Benoit,

Yes, please try it :)

It won't fix the pipeline config switching back to the primary if the context does change in some fundamental way but it should stop the issue you were having.

Let me know if it works or not...
Alan

File: python/tank/context.py [hunk 5]
```diff
@@ -113,23 +113,78 @@ def __str__(self):
...
+        if not self._entity_dicts_eq(self.step, other.step):
+            return False
+        
+        if not self._entity_dicts_eq(self.task, other.task):
+            return False
+        
+        if not self._entity_dicts_eq(self.user, other.user):
+            return False
+
+        # compare additional entities - note that currently this assumes the additional entity
```
Comment (jfboismenu-adsk) [inline python/tank/context.py]: How do you know that that this assumption is always going to be valid? From what I understand, the user can add any additional_entities into the context, so they might not be in the same order, right?

**Comment by jfboismenu-adsk** [comment]
A unit test would be nice.

**Comment by benoit-leveau** [comment]
Hi Alan,

I tried the fix, but for some reason it's using the studio core (which
doesn't have the fix) so it didn't work.

There's something really tricky to setup so that "import sgtk" will
correctly import the studio core (case #1), the per-project localized core
(case #2) or the sandbox localized core (case #3).
Add to that the fact that when running "tank Task xxx launch_maya" it
launches a new environment and it's becoming a bit of a nightmare.
I used to have case 1 and 3 working fine, and discovered case 2 wasn't
working, fixed it last week but apparently case 3 is not working anymore...

Benoit

**Comment by benoit-leveau** [comment]
Actually just after hitting Send, I managed to revert my previous fix for
the environment and I can use your fix. I confirm it indeed fixes the issue
when comparing the contexts. As you can see in the log below, even though
the name is different for the HumanUser, it still returns equality.

comparing '{'type': 'Project', 'id': 74, 'name': 'TankTest'}' and '{'type':
'Project', 'name': 'TankTest', 'id': 74}'
comparing '{'type': 'Asset', 'id': 1317, 'name': 'dalek'}' and '{'type':
'Asset', 'name': 'dalek', 'id': 1317}'
comparing '{'type': 'Step', 'id': 10, 'name': 'Modelling'}' and '{'type':
'Step', 'name': 'Modelling', 'id': 10}'
comparing '{'type': 'Task', 'name': 'Modelling', 'id': 19596}' and
'{'type': 'Task', 'name': 'Modelling', 'id': 19596}'
comparing '{'type': 'HumanUser', 'id': 52, 'name': '*benoit*'}' and
'{'type': 'HumanUser', 'id': 52, 'name': '*Benoit Leveau*'}'
Contexts are identical

File: python/tank/context.py (hunk 3 above)
Comment (alandann) [inline python/tank/context.py]: good spot - have fixed

File: python/tank/context.py (hunk 5 above)
Comment (alandann) [inline python/tank/context.py]: I don't! - I think I was trying to avoid an expensive compare but these are rarely used anyway so doing the compare properly is better

File: python/tank/context.py (hunk 4 above)
Comment (alandann) [inline python/tank/context.py]: Really good point - I think I was assuming the source of these dictionaries wasn't reliable but (a) that's not very Pythonic and (b) none of the rest of the code in this class bothers checking that the lists/dictionaries are well formed!  So I've removed the type checks both here and in _entity_dicts_eq

File: python/tank/context.py (hunk 1 above)
Comment (alandann) [inline python/tank/context.py]: me too!

File: python/tank/context.py (hunk 2 above)
Comment (alandann) [inline python/tank/context.py]: changed

**Comment by alandann** [comment]
Thanks for testing Benoit and glad it's working :)

**Comment by alandann** [comment]
Fixes done - Manne or JF - want to take one last look?

**Comment by jfboismenu-adsk** [comment]
Looks good. Thanks for the unit test.

**Comment by manneohrstrom** [comment]
looks awesome!
